docs/man/man1/dsched_info.1.rst
.. _man1-dsched_info:

dsched_info
===========

.. include_body

dsched_info |mdash| Display information about the DynaSched installation

SYNOPSIS
--------

``dsched_info [options]``


DESCRIPTION
-----------

``dsched_info`` provides detailed information about the DynaSched
installation. It can be useful for at least three common scenarios:

#. Checking local configuration and seeing how DynaSched was installed.

#. Submitting bug reports / help requests to the DynaSched community
   (see :doc:`Getting help </getting-help>`).

#. Seeing a list of installed DynaSched plugins and querying what MCA
   parameters they support.


OPTIONS
-------

``dsched_info`` accepts the following options:

* ``-h`` | ``--help <arg0>``: Show help message. If the optional
  argument is not provided, then a generalized help message similar
  to the information provided here is returned. If an argument is
  provided, then a more detailed help message for that specific
  command line option is returned.

* ``-v`` | ``--verbose``: Enable debug output.

* ``-V`` | ``--version``: Print version and exit.

* ``-a`` | ``--all``: Show all configuration options and MCA
  parameters.

* ``--arch``: Show architecture on which DynaSched was compiled.

* ``-c`` | ``--config``: Show configuration options

* ``--hostname``: Show the hostname on which DynaSched was configured
  and built.

* ``--internal``: Show internal MCA parameters (not meant to be
  modified by users).

* ``--param <arg0>:<arg1>,<arg2>``: Show MCA parameters.  The first
  parameter is the framework (or the keyword "all"); the second parameter
  is a comma-delimited list of specific component names (if only <arg0>
  is given, then all components will be reported).

* ``--path <type>``: Show paths that DynaSched was configured
  with. Accepts the following parameters: ``all``, ``prefix``, ``bindir``,
  ``libdir``, ``incdir``, ``pkglibdir``, ``sysconfdir``.

* ``--pretty-print``: When used in conjunction with other parameters, the output is
  displayed in "prettyprint" format (default)

* ``--parsable``: When used in conjunction with other parameters, the output is
  displayed in a machine-parsable format

* ``--parseable``: Synonym for ``--parsable``

* ``--color <hue>``: Control color coding: auto (default), never, always

* ``--show-failed``: Show the components that failed to load along with the reason why they failed

* ``--selected-only``: Show only variables from selected components.


EXIT STATUS
-----------

Returns 0 if successful, non-zero if an error is encountered

EXAMPLES
--------

Examples of using this command.

Show the default output of options and listing of installed
components in a human-readable / prettyprint format:

.. code-block::

   dsched_info

Show the default output of options and listing of installed components
in a machine-parsable format:

.. code-block::

   dsched_info --parsable

Show the MCA parameters of the "fifo" SCHED component in a
human-readable / prettyprint format:

.. code-block::

   dsched_info --param fifo sched

Show the "bindir" that DynaSched was configured with:

.. code-block::

   dsched_info --path bindir

Show the version of DynaSched version numbers in a prettyprint format:

.. code-block::

   dsched_info --version

Show *all* information about the DynaSched installation, including all
components that can be found, all the MCA parameters that they support,
versions of DynaSched and the components, etc.:

.. code-block::

   dsched_info --all

.. seealso::
   :ref:`dynasched(7) <man7-dynasched>`


docs/history.rst
History
========

Scheduling execution of the applications in such systems is often characterized by constraints on the available energy, communications bandwidth, and processing power associated with each node.

The Lagrangian approach to constrained problems was initially developed as a method for dealing with systems involving nonholonomic constraints – i.e., constraints that cannot be expressed as a function of the form
*f(r1, r2, r3, ..., t) = 0* [Go50]. Examples of nonholonomic
constraints include the walls of a container that constrain the motions of a gas, or a particle constrained to move on the surface of a sphere. In these situations, the equations governing the behavior of the system can be solved by introducing the constraints into those equations through the use of Lagrangian multipliers – undetermined parameters that are often functions of time.

Since that time, the Lagrangian technique has been applied to an ever-increasing range of problems that involve complex system equations and nonholonomic constraints [LeM67] [BrH75]. Extending the technique to the resource management problem where heuristics must be bound by “hard” constraints on precedence, energy, and time – all nonholonomic in nature – was a natural fit.
The most common use of the Lagrangian technique for scheduling has been in the manufacturing area. In these cases, the problem is to optimally schedule the manufacture of parts in an industrial operation where the parts require multiple machining operations and have desired completion times. There is a penalty for finishing too early or too late. The machining operations are con- strained by the available resources in terms of the types of machines used, the finite number of machines of each type, and the scheduling precedence requirements.

Luh and Hoitomt [LuH93] successfully adopted the Lagrangian approach by breaking the overall manufacturing problem into a series of sub-problems. This approach combines Lagrangian relaxation techniques with post- solution scheduling heuristics. The Lagrangian multipliers attempt to measure the marginal costs associated with machine capacity and scheduling precedence constraints. As dynamic changes in the manufacturing environment occur, the pre-existing optimal values of the Lagrangian multipliers can be used as a starting point for the computations for the changed environment. However, the resulting Lagrangian-based solutions typically represent infeasible schedules in that there are temporal machine capacity and scheduling precedence constraint violations. The final solution requires the use of list scheduling techniques whereby a greedy scheduling heuristic is used to sequentially schedule tasks by order of decreasing marginal costs. The problem is effectively treated as a scheduling solution to a static environment and the Lagrangian multipliers cannot be modified, and generate a guaranteed feasible solution, in a continuously changing environment.

Subsequently, Luh et al. [LuZ00] extended the Lagrangian technique through the introduction of a neural net- work that adjusted the Lagrangian multipliers as the scheduling progressed to achieve closer-to-optimal results. The key result in this work was a proof of the convergence of the Lagrangian relaxation neural network (LRNN) approach for constrained optimization problems. There are no requirements on the differentiability of the describing functions or the continuity of the decision variables – provided that the evolution of the decision variables leads the Lagrangian to a global minimization.
This approach works well for statically mapping the part manufacturing sequence in an industrial environment. However, it has two significant limitations:

(a) the LRNN formulation generally converges to a solution that may not be entirely feasible because of the difficulty in characterizing some constraints. As a result, there is a final step that readjusts the map- ping to ensure the final solution is feasible; and

(b) as implemented, the technique cannot respond to dynamic changes in the production line, such as the loss of a machine, without rescheduling the entire production sequence.


src/tools/dsched_info/dsched_info.c
/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
/*
 * Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
 *                         University Research and Technology
 *                         Corporation.  All rights reserved.
 * Copyright (c) 2004-2006 The University of Tennessee and The University
 *                         of Tennessee Research Foundation.  All rights
 *                         reserved.
 * Copyright (c) 2004-2007 High Performance Computing Center Stuttgart,
 *                         University of Stuttgart.  All rights reserved.
 * Copyright (c) 2004-2005 The Regents of the University of California.
 *                         All rights reserved.
 * Copyright (c) 2007-2020 Cisco Systems, Inc.  All rights reserved
 * Copyright (c) 2010-2016 Los Alamos National Security, LLC.
 *                         All rights reserved.
 * Copyright (c) 2014-2020 Intel, Inc.  All rights reserved.
 * Copyright (c) 2015      Research Organization for Information Science
 *                         and Technology (RIST). All rights reserved.
 * Copyright (c) 2021-2026 Nanook Consulting  All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 */

#include "dsched_config.h"

#include <ctype.h>
#include <stdio.h>
#include <string.h>
#ifdef HAVE_UNISTD_H
#    include <unistd.h>
#endif
#ifdef HAVE_NETDB_H
#    include <netdb.h>
#endif
#ifdef HAVE_SYS_PARAM_H
#    include <sys/param.h>
#endif
#include <errno.h>
#include <signal.h>

#include "src/class/pmix_object.h"
#include "src/class/pmix_pointer_array.h"
#include "src/mca/base/pmix_base.h"
#include "src/runtime/pmix_info_support.h"
#include "src/runtime/pmix_rte.h"
#include "src/util/pmix_argv.h"
#include "src/util/pmix_basename.h"
#include "src/util/pmix_cmd_line.h"
#include "src/util/pmix_path.h"
#include "src/util/pmix_show_help.h"

#include "src/include/dsched_constants.h"
#include "src/include/dsched_frameworks.h"
#include "src/include/dsched_internal_version.h"
#include "src/include/dsched_portable_platform_real.h"
#include "src/include/dsched_globals.h"
#include "src/mca/dinstalldirs/dinstalldirs.h"
#include "src/runtime/dsched_rte.h"
#include "src/util/dsched_cmd_line.h"
#include "src/util/dsched_error.h"

static int register_framework_params(pmix_pointer_array_t *component_map)
{
    int rc;

    /* Register mca/base parameters */
    if (PMIX_SUCCESS != pmix_mca_base_open(NULL)) {
        pmix_show_help("help-dsched_info.txt", "lib-call-fail", true, "pmix_mca_base_open",
                       __FILE__, __LINE__);
        return PMIX_ERROR;
    }

    /* Register the PMIX layer's MCA parameters */
    if (PMIX_SUCCESS != (rc = dsched_register_params())) {
        fprintf(stderr, "dsched_register_params failed\n");
        return rc;
    }

    return pmix_info_register_project_frameworks("dsched", dsched_frameworks, component_map);
}

/*
 * Public variables
 */

const char *dsched_info_type_all = "all";
const char *dsched_info_type_dsched = "dsched";
const char *dsched_info_type_base = "base";

// cmd line definition
static struct option dinfooptions[] = {
    PMIX_OPTION_SHORT_DEFINE(PMIX_CLI_HELP, PMIX_ARG_OPTIONAL, 'h'),
    PMIX_OPTION_SHORT_DEFINE(PMIX_CLI_VERSION, PMIX_ARG_NONE, 'V'),
    PMIX_OPTION_SHORT_DEFINE(PMIX_CLI_VERBOSE, PMIX_ARG_NONE, 'v'),

    // MCA parameters
    PMIX_OPTION_DEFINE(DSCHED_CLI_PMIXMCA, PMIX_ARG_REQD),
    PMIX_OPTION_DEFINE(DSCHED_CLI_DMCA, PMIX_ARG_REQD),
    PMIX_OPTION_DEFINE(DSCHED_CLI_MCA, PMIX_ARG_REQD),

    PMIX_OPTION_SHORT_DEFINE(PMIX_CLI_INFO_ALL, PMIX_ARG_NONE, 'a'),
    PMIX_OPTION_DEFINE(PMIX_CLI_INFO_ARCH, PMIX_ARG_NONE),
    PMIX_OPTION_SHORT_DEFINE(PMIX_CLI_INFO_CONFIG, PMIX_ARG_NONE, 'c'),
    PMIX_OPTION_DEFINE(PMIX_CLI_INFO_HOSTNAME, PMIX_ARG_NONE),
    PMIX_OPTION_DEFINE(PMIX_CLI_INFO_INTERNAL, PMIX_ARG_NONE),

    PMIX_OPTION_DEFINE(PMIX_CLI_INFO_PARAM, PMIX_ARG_REQD),
    PMIX_OPTION_DEFINE(PMIX_CLI_INFO_PARAMS, PMIX_ARG_REQD),
    PMIX_OPTION_DEFINE(PMIX_CLI_INFO_PATH, PMIX_ARG_REQD),
    PMIX_OPTION_DEFINE(PMIX_CLI_INFO_VERSION, PMIX_ARG_REQD),
    PMIX_OPTION_DEFINE(PMIX_CLI_PRETTY_PRINT, PMIX_ARG_NONE),
    PMIX_OPTION_DEFINE(PMIX_CLI_PARSABLE, PMIX_ARG_NONE),
    PMIX_OPTION_DEFINE(PMIX_CLI_PARSEABLE, PMIX_ARG_NONE),
    PMIX_OPTION_DEFINE(PMIX_CLI_INFO_SHOW_FAILED, PMIX_ARG_NONE),
    PMIX_OPTION_DEFINE(PMIX_CLI_INFO_SELECTED_ONLY, PMIX_ARG_NONE),
    PMIX_OPTION_SHORT_DEFINE(PMIX_CLI_INFO_TYPES, PMIX_ARG_REQD, 't'),
    PMIX_OPTION_DEFINE(PMIX_CLI_INFO_COLOR, PMIX_ARG_REQD),

    PMIX_OPTION_DEFINE("include-pmix", PMIX_ARG_NONE),

    /* End of list */
    PMIX_OPTION_END
};
static char *dinfoshorts = "h::vVact:";


int main(int argc, char *argv[])
{
    int ret = 0;
    pmix_status_t rc;
    bool acted = false;
    bool want_all = false;
    bool include_pmix = false;
    int i;
    char *color;
    char *str;
    char *ptr;
    pmix_pointer_array_t dsched_component_map;
    pmix_pointer_array_t mca_types;
    pmix_cli_result_t results;
    pmix_cli_item_t *opt;
    DSCHED_HIDE_UNUSED_PARAMS(argc);

    /* protect against problems if someone passes us thru a pipe
     * and then abnormally terminates the pipe early */
    signal(SIGPIPE, SIG_IGN);

    /* Initialize the argv parsing stuff */
    dsched_globals.basename = pmix_basename(argv[0]);
    if (DSCHED_SUCCESS != (ret = dsched_init_util())) {
        pmix_show_help("help-dsched_info.txt", "lib-call-fail", true, "dsched_init_util", __FILE__,
                       __LINE__, NULL);
        exit(ret);
    }

    /* parse the input argv to get values, including everyone's MCA params */
    PMIX_CONSTRUCT(&results, pmix_cli_result_t);
    rc = pmix_cmd_line_parse(argv, dinfoshorts, dinfooptions, NULL,
                             &results, "help-dsched_info.txt");
    if (PMIX_SUCCESS != rc) {
        PMIX_DESTRUCT(&results);
        if (PMIX_OPERATION_SUCCEEDED == rc) {
            return PMIX_SUCCESS;
        }
        if (PMIX_ERR_SILENT != rc) {
            fprintf(stderr, "%s: command line error (%s)\n", dsched_globals.basename, PMIx_Error_string(rc));
        }
        return rc;
    }

    // we do NOT accept arguments other than our own
    if (NULL != results.tail) {
        str = PMIx_Argv_join(results.tail, ' ');
        if (0 != strcmp(str, argv[0])) {
            ptr = pmix_show_help_string("help-dsched_info.txt", "no-args", false,
                                        dsched_globals.basename, str, dsched_globals.basename);
            free(str);
            if (NULL != ptr) {
                printf("%s", ptr);
                free(ptr);
            }
            return -1;
        }
        free(str);
    }

    // see if they want PMIx info included
    include_pmix = pmix_cmd_line_is_taken(&results, "include-pmix");

    /* Determine color support */
    opt = pmix_cmd_line_get_param(&results, PMIX_CLI_INFO_COLOR);
    if (NULL != opt) {
        color = NULL;
        if (NULL != opt->values) {
            color = opt->values[0];
        }
        if (NULL == color) {
            color = "auto";
        }
    } else {
        color = "auto";
    }
    if (0 == strcasecmp(color, "auto")) {
        #if HAVE_ISATTY
            pmix_info_color = isatty(STDOUT_FILENO);
        #else
            pmix_info_color = false;
        #endif
    } else if (0 == strcasecmp(color, "always")) {
        pmix_info_color = true;
    } else if (0 == strcasecmp(color, "never")) {
        pmix_info_color = false;
    } else {
        fprintf(stderr, "%s: Unrecognized value '%s' to color parameter\n", argv[0], color);
        exit(2);
    }

    /* set the flags */
    if (pmix_cmd_line_is_taken(&results, PMIX_CLI_PRETTY_PRINT)) {
        pmix_info_pretty = true;
    } else if (pmix_cmd_line_is_taken(&results, PMIX_CLI_PARSABLE) ||
               pmix_cmd_line_is_taken(&results, PMIX_CLI_PARSEABLE)) {
        pmix_info_pretty = false;
    }

    if (pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_SELECTED_ONLY)) {
        /* register only selected components */
        pmix_info_register_flags = PMIX_MCA_BASE_REGISTER_DEFAULT;
    }

    if (pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_SHOW_FAILED)) {
        pmix_mca_base_component_track_load_errors = true;
    }

    /* setup the mca_types array */
    PMIX_CONSTRUCT(&mca_types, pmix_pointer_array_t);
    pmix_pointer_array_init(&mca_types, 256, INT_MAX, 128);

    /* add a type for dsched itself */
    pmix_pointer_array_add(&mca_types, "mca");
    pmix_pointer_array_add(&mca_types, "dsched");

    /* push all the types found by autogen */
    for (i = 0; NULL != dsched_frameworks[i]; i++) {
        pmix_pointer_array_add(&mca_types, dsched_frameworks[i]->framework_name);
    }

    if (include_pmix) {
        // push all the PMIx framework types
        pmix_info_register_types(&mca_types, true);
    }

    /* setup the component_map array */
    PMIX_CONSTRUCT(&dsched_component_map, pmix_pointer_array_t);
    pmix_pointer_array_init(&dsched_component_map, 256, INT_MAX, 128);

    /* Register all global MCA Params */
    if (DSCHED_SUCCESS != (ret = dsched_register_params())) {
        if (DSCHED_ERR_SILENT != ret) {
            pmix_show_help("help-dsched-runtime", "dsched_init:startup:internal-failure", true,
                           "dsched register params",
                           DSCHED_ERROR_NAME(ret), ret);
        }
        return 1;
    }

     /* Register framework/component params */
    if (PMIX_SUCCESS != (ret = register_framework_params(&dsched_component_map))) {
        if (PMIX_ERR_BAD_PARAM == ret) {
            /* output what we got */
            pmix_info_do_params("DynaSched", true, pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_INTERNAL),
                                &mca_types, &dsched_component_map, NULL);
        }
        exit(1);
    }

    if (include_pmix) {
        /* register params for pmix */
        if (PMIX_SUCCESS != (ret = pmix_register_params())) {
            fprintf(stderr, "pmix_register_params failed with %d\n", ret);
            return PMIX_ERROR;
        }


        /* Register PMIx's params */
        if (PMIX_SUCCESS != (ret = pmix_info_register_framework_params(&dsched_component_map))) {
            if (PMIX_ERR_BAD_PARAM == ret) {
                /* output what we got */
                pmix_info_do_params("PMIx", true, pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_INTERNAL),
                                    &mca_types, &dsched_component_map, NULL);
            }
            exit(1);
        }
    }

    /* Execute the desired action(s) */
    want_all = pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_ALL);

    opt = pmix_cmd_line_get_param(&results, PMIX_CLI_INFO_VERSION);
    if (want_all || NULL != opt) {
        if (NULL == opt) {
            pmix_info_show_package(DSCHED_PACKAGE_STRING);
            pmix_info_show_version("DynaSched", pmix_info_ver_full, DSCHED_MAJOR_VERSION, DSCHED_MINOR_VERSION,
                                   DSCHED_RELEASE_VERSION, DSCHED_GREEK_VERSION, DSCHED_REPO_REV,
                                   DSCHED_RELEASE_DATE);
            pmix_info_show_component_version("DynaSched", &mca_types, &dsched_component_map, pmix_info_type_all,
                                             pmix_info_component_all, pmix_info_ver_full,
                                             pmix_info_ver_all);

        } else if (NULL == opt->values) {
            pmix_info_show_package(DSCHED_PACKAGE_STRING);
            pmix_info_show_version("DynaSched", pmix_info_ver_full, DSCHED_MAJOR_VERSION, DSCHED_MINOR_VERSION,
                                    DSCHED_RELEASE_VERSION, DSCHED_GREEK_VERSION, DSCHED_REPO_REV,
                                    DSCHED_RELEASE_DATE);
            pmix_info_show_component_version("DynaSched", &mca_types, &dsched_component_map, pmix_info_type_all,
                                             pmix_info_component_all, pmix_info_ver_full,
                                             pmix_info_ver_all);

        } else {
            if (0 == strcasecmp(opt->values[0], "dsched") ||
                0 == strcasecmp(opt->values[0], "all")) {
                pmix_info_show_package(DSCHED_PACKAGE_STRING);
                pmix_info_show_version("DynaSched", pmix_info_ver_full, DSCHED_MAJOR_VERSION, DSCHED_MINOR_VERSION,
                                       DSCHED_RELEASE_VERSION, DSCHED_GREEK_VERSION, DSCHED_REPO_REV,
                                       DSCHED_RELEASE_DATE);
                pmix_info_show_component_version("DynaSched", &mca_types, &dsched_component_map, pmix_info_type_all,
                                                 pmix_info_component_all, pmix_info_ver_full,
                                                 pmix_info_ver_all);

            } else {
                // the first arg is either the name of a framework, or a framework:component pair
                char **tmp = PMIx_Argv_split(opt->values[0], ':');
                const char *component = pmix_info_component_all;
                const char *modifier = pmix_info_ver_full;
                if (NULL != tmp[1]) {
                    component = tmp[1];
                }
                if (NULL != opt->values[1]) {
                    modifier = opt->values[1];
                }
                pmix_info_show_component_version("DynaSched", &mca_types, &dsched_component_map, tmp[0],
                                                 component, modifier,
                                                 pmix_info_ver_all);
                PMIx_Argv_free(tmp);
            }
            acted = true;
        }
    }

    if (want_all || pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_PATH)) {
        pmix_info_do_path(want_all, &results, &dsched_dinstall_dirs);
        acted = true;
    }

    if (want_all || pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_ARCH)) {
        pmix_info_do_arch();
        acted = true;
    }

    if (!want_all && pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_HOSTNAME)) {
        // hostname is contain in do_config, so don't duplicate it here
        pmix_info_do_hostname();
        acted = true;
    }

    if (want_all || pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_CONFIG)) {
        pmix_info_do_config(true, DSCHED_CONFIGURE_USER, DSCHED_CONFIGURE_DATE, DSCHED_CONFIGURE_HOST,
                            DSCHED_CONFIGURE_CLI, DSCHED_BUILD_USER, DSCHED_BUILD_DATE, DSCHED_BUILD_HOST,
                            DSCHED_CC, DSCHED_CC_ABSOLUTE, PLATFORM_STRINGIFY(PLATFORM_COMPILER_FAMILYNAME),
                            PLATFORM_STRINGIFY(PLATFORM_COMPILER_VERSION_STR), DSCHED_BUILD_CFLAGS, DSCHED_BUILD_LDFLAGS,
                            DSCHED_BUILD_LIBS, DSCHED_ENABLE_DEBUG, DSCHED_HAVE_DDL_SUPPORT,
                            DSCHED_C_HAVE_VISIBILITY);
        acted = true;
    }

    if (want_all || pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_PARAM) ||
        pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_PARAMS)) {
        pmix_info_do_params("DynaSched", true, pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_INTERNAL),
                            &mca_types, &dsched_component_map, &results);
        acted = true;
    }

    if (pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_TYPES)) {
        pmix_info_do_type(&results);
        acted = true;
    }

    /* If no command line args are specified, show default set */

    if (!acted) {
        pmix_info_show_package(DSCHED_PACKAGE_STRING);
        pmix_info_show_version("DynaSched", pmix_info_ver_full, DSCHED_MAJOR_VERSION, DSCHED_MINOR_VERSION,
                               DSCHED_RELEASE_VERSION, DSCHED_GREEK_VERSION, DSCHED_REPO_REV,
                               DSCHED_RELEASE_DATE);
        pmix_info_show_path(pmix_info_path_prefix, dsched_dinstall_dirs.prefix);
        pmix_info_do_arch();
        pmix_info_do_config(false, DSCHED_CONFIGURE_USER, DSCHED_CONFIGURE_DATE, DSCHED_CONFIGURE_HOST,
                            DSCHED_CONFIGURE_CLI, DSCHED_BUILD_USER, DSCHED_BUILD_DATE, DSCHED_BUILD_HOST,
                            DSCHED_CC, DSCHED_CC_ABSOLUTE, PLATFORM_STRINGIFY(PLATFORM_COMPILER_FAMILYNAME),
                            PLATFORM_STRINGIFY(PLATFORM_COMPILER_VERSION_STR), DSCHED_BUILD_CFLAGS, DSCHED_BUILD_LDFLAGS,
                            DSCHED_BUILD_LIBS, DSCHED_ENABLE_DEBUG, DSCHED_HAVE_DDL_SUPPORT,
                            DSCHED_C_HAVE_VISIBILITY);
        pmix_info_show_component_version("DynaSched", &mca_types, &dsched_component_map, pmix_info_type_all,
                                         pmix_info_component_all, pmix_info_ver_full,
                                         pmix_info_ver_all);
    }

    // if requested, do the PMIx info
    if (include_pmix) {
        acted = false;
        pmix_info_out(NULL, NULL, "\n\n\n==================");
        pmix_info_out(NULL, NULL, "\nPMIx CONFIGURATION\n");
        if (want_all || NULL != opt) {
            if (NULL == opt) {
                pmix_info_show_pmix_package();
                pmix_info_show_pmix_version();
                pmix_info_show_component_version("PMIx", &mca_types, &dsched_component_map, pmix_info_type_all,
                                                 pmix_info_component_all, pmix_info_ver_full,
                                                 pmix_info_ver_all);

            } else if (NULL == opt->values) {
                pmix_info_show_pmix_package();
                pmix_info_show_pmix_version();
                pmix_info_show_component_version("PMIx", &mca_types, &dsched_component_map, pmix_info_type_all,
                                                 pmix_info_component_all, pmix_info_ver_full,
                                                 pmix_info_ver_all);
                if (include_pmix) {
                    pmix_info_show_component_version("PMIx", &mca_types, &dsched_component_map, pmix_info_type_all,
                                                     pmix_info_component_all, pmix_info_ver_full,
                                                     pmix_info_ver_all);
                }

            } else {
                if (0 == strcasecmp(opt->values[0], "pmix") ||
                    0 == strcasecmp(opt->values[0], "all")) {
                    pmix_info_show_pmix_package();
                    pmix_info_show_pmix_version();
                    pmix_info_show_component_version("PMIx", &mca_types, &dsched_component_map, pmix_info_type_all,
                                                     pmix_info_component_all, pmix_info_ver_full,
                                                     pmix_info_ver_all);

                } else {
                    // the first arg is either the name of a framework, or a framework:component pair
                    char **tmp = PMIx_Argv_split(opt->values[0], ':');
                    const char *component = pmix_info_component_all;
                    const char *modifier = pmix_info_ver_full;
                    if (NULL != tmp[1]) {
                        component = tmp[1];
                    }
                    if (NULL != opt->values[1]) {
                        modifier = opt->values[1];
                    }
                    pmix_info_show_component_version("PMIx", &mca_types, &dsched_component_map, tmp[0],
                                                     component, modifier,
                                                     pmix_info_ver_all);

                    PMIx_Argv_free(tmp);
                }
                acted = true;
            }
        }

        if (want_all || pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_PATH)) {
            pmix_info_do_pmix_path(want_all, &results);
            acted = true;
        }

        if (want_all || pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_ARCH)) {
            pmix_info_do_arch();
            acted = true;
        }

        if (!want_all && pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_HOSTNAME)) {
            // hostname is contain in do_config, so don't duplicate it here
            pmix_info_do_hostname();
            acted = true;
        }

        if (want_all || pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_CONFIG)) {
            pmix_info_do_pmix_config(want_all);
            acted = true;
        }

        if (want_all || pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_PARAM) ||
            pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_PARAMS)) {
             pmix_info_do_params("PMIx", true, pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_INTERNAL),
                                &mca_types, &dsched_component_map, &results);
            acted = true;
        }

        if (pmix_cmd_line_is_taken(&results, PMIX_CLI_INFO_TYPES)) {
            pmix_info_do_type(&results);
            acted = true;
        }

        /* If no command line args are specified, show default set */

    if (!acted) {
            pmix_info_show_pmix_package();
            pmix_info_show_pmix_version();
            pmix_info_show_path(pmix_info_path_prefix, pmix_pinstall_dirs.prefix);
            pmix_info_do_arch();
            pmix_info_do_pmix_config(want_all);
            pmix_info_show_component_version("PMIx", &mca_types, &dsched_component_map, pmix_info_type_all,
                                             pmix_info_component_all, pmix_info_ver_full,
                                             pmix_info_ver_all);
        }
    }
    /* All done */
    PMIX_DESTRUCT(&mca_types);
    pmix_mca_base_close();

    return 0;
}
